### PR TITLE
Utilize Indices for Key Lookup and Filtering Attestations in DB

### PR DESCRIPTION
**Pull request by rauljordan**
Part of #3064 

---

# Description

**Write why you are making the changes in this pull request**

Previously, our approaches for filtering and fetching information from our key-value store were all extremely inefficient, often looping through the entire set of key and values for a type just to find a certain element by a byte-prefixed key. For example, boltDB provides these utilities for performing prefix scans:

```go
db.View(func(tx *bolt.Tx) error {
	// Assume bucket exists and has keys
	c := tx.Bucket([]byte("MyBucket")).Cursor()

	prefix := []byte("1234")
	for k, v := c.Seek(prefix); k != nil && bytes.HasPrefix(k, prefix); k, v = c.Next() {
		fmt.Printf("key=%s, value=%s\n", k, v)
	}

	return nil
})
```

The seek function underneath the hood can be as fast as binary search, but we argue we can still do better if we pursue a different approach.

**Write a summary of the changes you are making**

Instead of any sort of scan across a DB bucket, we follow an approach of tracking certain attributes as indices for the different types we care to store in the DB. For attestations, for example, we care about filtering them by their shard number, their parent root, start epoch, end epoch, among other fields. That is, what we can do is keep indices such as `shard-5` that store a list of **attestation data roots** for that particular shard. Then, when we perform a lookup over the attestations that are from shard 5, we simply construct that index, fetch the roots stored at that index, and then do a batch read using those roots to retrieve all matching attestations. This is a very fast lookup once we have the roots, as there is no looping required over the keys nor values in the store just to find a piece of data. The downside is these indices need to be updated every time we save/delete attestations from the DB.

For multiple filter criteria, such as wanting to fetch all attestations with shard 5 and parent root `0xf00`, we would construct indices `shard-5` and `0xf00` and then proceed to fetch the attestation roots stored at those indices. We would then need to compute the intersection of these roots, and use that result to lookup all matching attestations.

We believe this approach will scale better over time and as the database grows in production.

**Link anything that would be helpful or relevant to the reviewers**

The keys for looking up attestations are `HashTreeRoot(attestation.Data)`, as this would allow us to treat aggregated attestations the same in lookups as non-aggregated ones with the same data.

### CODE REVIEW

File: beacon-chain/db/kv/attestations.go [hunk 1]
```diff
@@ -1,89 +1,113 @@
 package kv
```
Comment (rauljordan) [inline beacon-chain/db/kv/attestations.go]: Hey team, please review this file carefully, as most of this logic will be reused for blocks as well.

File: beacon-chain/db/kv/attestations.go [hunk 2]
```diff
@@ -106,81 +135,67 @@ func (k *Store) SaveAttestations(ctx context.Context, atts []*ethpb.Attestation)
...
-				numCriteriaMet++
-			}
-		case filters.Shard:
-			if bytes.Contains(key, append([]byte("shard"), uint64ToBytes(v.(uint64))...)) {
-				numCriteriaMet++
-			}
+			idx := append(endEpochIdx, uint64ToBytes(v.(uint64))...)
+			keys = append(keys, idx)
+		default:
+			return nil, fmt.Errorf("filter criterion %v not supported for attestations", k)
```
Comment (rauljordan) [inline beacon-chain/db/kv/attestations.go]: We directly specify the allowable criteria for filtering attestations. We wouldn't want a user trying to call `db.Attestations(ctx, filters.NewFilter().SetValidatorIndex(5))` when attestations do not have a validator index field for filtering. This is a good compromise that encourages good use of the library and prevents nasty errors/panics.

File: beacon-chain/db/kv/attestations.go [hunk 3]
```diff
@@ -106,81 +135,67 @@ func (k *Store) SaveAttestations(ctx context.Context, atts []*ethpb.Attestation)
...
 			return err
 		}
 		encodedValues[i] = enc
-		keys[i] = key
+		keys[i] = key[:]
 	}
 	return k.db.Batch(func(tx *bolt.Tx) error {
-		bucket := tx.Bucket(attestationsBucket)
+		bkt := tx.Bucket(attestationsBucket)
 		for i := 0; i < len(atts); i++ {
```
Comment (prestonvanloon) [inline beacon-chain/db/kv/attestations.go]: Would it be faster to have one loop over the attestations? It seems that you loop over it twice

Comment (rauljordan) [reply beacon-chain/db/kv/attestations.go]: The alternative is to do the hash tree roots inside of the db tx, which is an expensive operation given the tx acquires a db lock

**Pull request merged**